include/lbann/layers/transform/upsample.hpp

```diff
@@ -0,0 +1,297 @@
+class upsample_layer : public data_type_layer<TensorDataType>
+{
+  static_assert(T_layout == data_layout::DATA_PARALLEL,
+                "upsample only supports DATA_PARALLEL");
+
+private:
+  /** Upsample mode. */
+  upsample_mode m_upsample_mode;
+
+  /** Output scale factors. */
+  std::vector<int> m_scale_factors;
+
+#ifdef LBANN_HAS_DNN_LIB
+  /** Pooling descriptor. */
+  dnn_lib::PoolingDescriptor m_pooling_dnn_desc;
+  /** Tensor DNN library descriptors. */
+  dnn_lib::data_parallel_layer_tensor_manager<TensorDataType>
+    m_tensors_dnn_desc;
+#endif // LBANN_HAS_DNN_LIB
+
+public:
+  upsample_layer(lbann_comm* comm,
+                int num_data_dims,
+                int scale_factors,
+                upsample_mode mode)
+    : upsample_layer(comm,
+                    num_data_dims,
+                    std::vector<int>(num_data_dims, scale_factors),
+                    mode)
+  {}
+
+  upsample_layer(lbann_comm* comm,
+                int num_data_dims,
+                std::vector<int> scale_factors,
+                upsample_mode mode)
+    : data_type_layer<TensorDataType>(comm),
+      m_upsample_mode(mode),
+      m_scale_factors{std::move(scale_factors)}
+#ifdef LBANN_HAS_DNN_LIB
+      ,
+      m_tensors_dnn_desc(this)
+#endif // LBANN_HAS_DNN_LIB
+  {}
+
+  upsample_layer(const upsample_layer& other)
+    : data_type_layer<TensorDataType>(other),
+      m_upsample_mode(other.m_upsample_mode),
+      m_scale_factors(other.m_scale_factors)
+#ifdef LBANN_HAS_DNN_LIB
+      ,
+      m_pooling_dnn_desc(other.m_pooling_dnn_desc),
+      m_tensors_dnn_desc(other.m_tensors_dnn_desc)
+#endif // LBANN_HAS_DNN_LIB
+  {
+#ifdef LBANN_HAS_DNN_LIB
+    m_tensors_dnn_desc.set_layer(this);
+#endif // LBANN_HAS_DNN_LIB
+  }
+
+  upsample_layer& operator=(const upsample_layer& other)
+  {
+    data_type_layer<TensorDataType>::operator=(other);
+    m_upsample_mode = other.m_upsample_mode;
+    m_scale_factors = other.m_scale_factors;
+#ifdef LBANN_HAS_DNN_LIB
+    m_pooling_dnn_desc = other.m_pooling_dnn_desc;
+    m_tensors_dnn_desc = other.m_tensors_dnn_desc;
+    m_tensors_dnn_desc.set_layer(this);
+#endif // LBANN_HAS_DNN_LIB
+    return *this;
+  }
+
+  ~upsample_layer() override = default;
+
+  upsample_layer* copy() const override { return new upsample_layer(*this); }
+
+  /** @name Serialization */
+  ///@{
+
+  template <typename ArchiveT>
+  void serialize(ArchiveT& ar);
+
+  ///@}
+
+  std::string get_type() const override { return "upsample"; }
+  data_layout get_data_layout() const override { return T_layout; }
+  El::Device get_device_allocation() const override { return Dev; }
+
+#ifdef LBANN_HAS_ONNX
+  void fill_onnx_node(onnx::GraphProto& graph) const override;
+#endif // LBANN_HAS_ONNX
+
+  description get_description() const override
+  {
+    auto desc = data_type_layer<TensorDataType>::get_description();
+    std::stringstream ss;
+
+    // Upsample mode
+    ss.str(std::string{});
+    ss.clear();
```

Comment:
I don't think these two lines are necessary. Also, this could just be a std::string or the switch-case can call desc.add(...) directly.

Reply:
fixed